source/libraries/ContractInterfacesGenerator.ts

```diff
@@ -0,0 +1,191 @@
+
+        const contractFunctions: Array<AbiFunction> = contractAbi
+            .filter(v => v.type == "function")
+            .map(v => <AbiFunction>v);
+
+        for (let abiFunction of contractFunctions) {
+            if (seen.has(abiFunction.name)) continue;
+            if (!abiFunction.constant) {
+                contractMethods.push(this.remoteMethodTemplate(abiFunction));
+            }
+            contractMethods.push(this.localMethodTemplate(abiFunction));
+            seen.add(abiFunction.name);
+        }
+
+        let extendsControlled: boolean = seen.has("getController") && contractName != "Controlled";
+
+        return `export class ${contractName} extends ${extendsControlled ? "Controlled" : "Contract"} {
+    public constructor(connector: Connector, accountManager: AccountManager, address: string, defaultGasPrice: BN) {
+        super(connector, accountManager, address, defaultGasPrice);
+    }
+
+${contractMethods.join("\n\n")}
+}
+`
+    }
+
+    private remoteMethodTemplate(abiFunction: AbiFunction) {
+        let argNames: String = this.getArgNamesString(abiFunction);
+        let params: String = this.getParamsString(abiFunction);
+        let options: String = `{ sender?: string, gasPrice?: BN${abiFunction.payable ? ", attachedEth?: BN" : ""} }`;
+        return `    public ${abiFunction.name} = async(${params} options?: ${options}): Promise<string> => {
+        options = options || {};
+        const abi: AbiFunction = ${JSON.stringify(abiFunction)};
+        return await this.remoteCall(abi, [${argNames}], options.sender, options.gasPrice${abiFunction.payable ? ", options.attachedEth" : ""});
+    }`;
+    }
+
+    private localMethodTemplate(abiFunction: AbiFunction) {
+        let argNames: String = this.getArgNamesString(abiFunction);
+        let params: String = this.getParamsString(abiFunction);
+        let options: String = `{ sender?: string${abiFunction.payable ? ", attachedEth?: BN" : ""} }`;
+        let returnType: String = this.getTsTypeFromPrimitive(abiFunction.outputs[0].type, true);
+        let returnPromiseType: String = abiFunction.outputs.length == 1 ? returnType : "Array<string>";
+        let returnValue: String = abiFunction.outputs.length == 1 ? `<${returnType}>result[0]` : "<Array<string>>result";
+        return `    public ${abiFunction.name}_ = async(${params} options?: ${options}): Promise<${returnPromiseType}> => {
+        options = options || {};
+        const abi: AbiFunction = ${JSON.stringify(abiFunction)};
+        const result = await this.localCall(abi, [${argNames}], options.sender${abiFunction.payable ? ", options.attachedEth" : ""});
+        return ${returnValue};
+    }`;
+    }
+
+    private getTsTypeFromPrimitive(abiType: Primitive, forReturnValue?: boolean) {
+        switch(abiType) {
+            case 'uint8':
+            case 'uint64':
+            case 'uint256':
+            case 'int256': {
+                return forReturnValue ? 'BN' : 'BN|number';
```

Comment:
I did this (BN|number) to make the interfaces easier to use.  However, it does allow the user to pass in a floating point number which causes problems.  We _could_ require the user always pass in a BN instead, just makes some call sites ugly when all you want to do is pass in `1` or something.

Reply:
That would simplify things and make this more correct so I'm in favor if you think that's alright.

Reply:
To clarify I mean in favor of making this just `BN`